### PR TITLE
Applies zone labels to newly created vsphere volumes

### LINKED ISSUE #73301: vSphere volumes do not have zone labels auto applied
**What happened**:
With zones marked using tags in vCenter and using VCP, the k8s nodes show the zone information correctly in its labels, but a volume does not have the zone labels applied. A node's zone information works correctly due to PR https://github.com/kubernetes/kubernetes/pull/66795.

```
$:~> kubectl get nodes --show-labels
NAME                 STATUS                     AGE       VERSION                                     LABELS
kubernetes2-master   Ready,SchedulingDisabled   6d        v1.14.0-alpha.0.1570+ccf84273118a1f-dirty   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=region_1,failure-domain.beta.kubernetes.io/zone=zone_1,kubernetes.io/hostname=kubernetes2-master
kubernetes2-node1    Ready                      6d        v1.14.0-alpha.0.1570+ccf84273118a1f-dirty   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=region_1,failure-domain.beta.kubernetes.io/zone=zone_1,kubernetes.io/hostname=kubernetes2-node1
```

However, a newly created volume does not show any zone labels today.

```
$:~> kubectl get pv --show-labels
NAME                                       CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS    CLAIM          STORAGECLASS   REASON    AGE       LABELS
pvc-aab193fe-1fc8-11e9-a263-005056929afb   1Mi        RWO           Delete          Bound     default/pvc1   sc1                      17h       None
```

**What you expected to happen**:

A newly created volume should be automatically applied with the well-known zone labels.

```
$:~> kubectl get pv --show-labels
NAME                                       CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS    CLAIM          STORAGECLASS   REASON    AGE       LABELS
pvc-aab193fe-1fc8-11e9-a263-005056929afb   1Mi        RWO           Delete          Bound     default/pvc1   sc1                      17h       failure-domain.beta.kubernetes.io/region=region_1,failure-domain.beta.kubernetes.io/zone=zone_1
```

This is important since the scheduler places the pod on a node of the same zone based on the Volume's zone information.

**How to reproduce it (as minimally and precisely as possible)**:
1. Setup zones in vCenter by tagging an ESX host, cluster or datacenter with the tag name of your choice and tag category, say "zoneCategory" and "regionCategory".
2. Update the /etc/kubernetes/vsphere.conf file in each k8s node to include the following and restart kubelet.
```
[Labels]
    zone = "zoneCategory"
    region = "regionCategory"
```
3. Create a PV and look at the auto-applied labels on it as shown above. The zone information is missing.

**Anything else we need to know?**:
The partially addresses https://github.com/kubernetes/kubernetes/issues/67703

**Environment**:
- Kubernetes version (use `kubectl version`): any
- Cloud provider or hardware configuration: vSphere

### PR DESCRIPTION
**Pull request by subramanian-neelakantan**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When a volume is created on vSphere, the zone label does not get applied on it today. Without this information on the volume, a pod that uses this volume could get scheduled to a different zone where this volume is not visible/accessible. This issue can be solved by just labeling the volume with the zone information. This PR computes the zone of a volume in the PersistentVolumeLabel admission controller since vsphere is still an in-tree cloud provider and the out-of-tree plugin in still the works.

**Which issue(s) this PR fixes**:
Fixes #73301 
Part of issue kubernetes#67703

**Special notes for your reviewer**:
This part does not guarantee that allowedTopologies specified in a StorageClass will be honoured. That would be done as part of a different PR.

**Does this PR introduce a user-facing change?**:
Yes
```release-note
This change applies zone labels to vSphere Volumes automatically. The zone labels are visible on the PV:
$ kubectl get pv --show-labels
NAME           CAPACITY   ACCESSMODES   STATUS    CLAIM            REASON    AGE       LABELS
pv-abc                5Gi        RWO                   Bound     default/claim1                    46s       failure-domain.beta.kubernetes.io/region=VC1,failure-domain.beta.kubernetes.io/zone=cluster-1
```

### CODE REVIEW

File: pkg/cloudprovider/providers/vsphere/vclib/datastore.go [hunk 1]
```diff
@@ -85,3 +85,19 @@ func (ds *Datastore) IsCompatibleWithStoragePolicy(ctx context.Context, storageP
...
+// GetDatastoreHostMounts gets the host names mounted on given datastore
+func (ds *Datastore) GetDatastoreHostMounts(ctx context.Context) ([]types.ManagedObjectReference, error) {
+	var dsMo mo.Datastore
+	pc := property.DefaultCollector(ds.Client())
+	err := pc.RetrieveOne(ctx, ds.Datastore.Reference(), []string{"host"}, &dsMo)
+	if err != nil {
+		klog.Errorf("Failed to retrieve datastore host mount property. err: %v", err)
+		return nil, err
+	}
+	hosts := make([]types.ManagedObjectReference, 0)
```
Comment (SandeepPissay) [inline pkg/cloudprovider/providers/vsphere/vclib/datastore.go]:
```suggestion
	hosts := make([]types.ManagedObjectReference, len(dsMo.Host))
```

**Comment by SandeepPissay** [comment]
@subramanian-neelakantan Can you describe how you tested this PR?

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -1392,12 +1409,12 @@ func (vs *VSphere) GetZone(ctx context.Context) (cloudprovider.Zone, error) {
 		}
 
 		if zone.Region == "" {
-			if vs.cfg.Labels.Region != "" {
+			if vs.cfg.Labels.Region != DefaultRegionCategoryName {
```
Comment (SandeepPissay) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: GetZone method seems to assume that the k8s cluster is deployed in one single VC. So, zones would not work if the k8s cluster is deployed across multiple VCs? Note that VCP is designed to work when the k8s cluster is deployed across multiple VCs. 

One question - is zones feature optional? If so, how can one run VCP without zones support?

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 3]
```diff
@@ -1418,3 +1435,248 @@ func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.Node
 func (vs *VSphere) GetZoneByProviderID(ctx context.Context, providerID string) (cloudprovider.Zone, error) {
 	return cloudprovider.Zone{}, cloudprovider.NotImplemented
 }
+
+// GetVolumeLabels returns the well known zone and region labels for given volume
+func (vs *VSphere) GetVolumeLabels(volumePath string) (map[string]string, error) {
```
Comment (SandeepPissay) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Which datastore is available in which zone/region is going to be static. Any specific reasons for computing this information every time?

**Comment by subramanian-neelakantan** [comment]
> @subramanian-neelakantan Can you describe how you tested this PR?

This was tested by the e2e tests that @sandeeppsunny has written here https://github.com/kubernetes/kubernetes/pull/72732. The current PR along with @skarthiksrinivas PR https://github.com/kubernetes/kubernetes/pull/72731 that places volume on a datastore in a given zone were merged to run the e2e tests and validate the functionality.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 2 above)
Comment (subramanian-neelakantan) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: Multi-VC has not been tested with this change.

The zones feature is optional to use. When there are no zone tags in vCenter, VCP works as it did before this change.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 3 above)
Comment (subramanian-neelakantan) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: When creating a volume it is the admission controller that labels the volume with the zone information. The admission controller runs in the api-server and creates a new instance of a cloudprovider to fetch the zone information each time. So the cloudprovider cannot cache the zone information for a datastore.

A datastore is said to belong a zone if it is mounted on all ESX hosts of that zone. By that definition a datastore can belong to multiple zones. To calculate the zones of a datastore we do the following:
1. Compute the map of zone to all hosts in that zone. This is done across all ESX hosts in VC.
2. Find all the hosts to which the given datastore is mounted.
3. For each zone, if all the hosts in that zone are contained in the list of (2) - meaning the datastore is mounted on all hosts in that zone, then the zone is included.

**Comment by andrewsykim** [comment]
Heads up that there will be some conflicting changes if #73102 merges first :)

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 3 above)
Comment (andrewsykim) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Should use `GetLabelsForVolume` as per https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L232-L235. Other providers will do the same once https://github.com/kubernetes/kubernetes/pull/73102 is merged

Comment (subramanian-neelakantan) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: Done.

**Comment by vladimirvivien** [comment]
@subramanian-neelakantan 
Since this is a bug, please create an issue explaining the problem.  Then refer this issue in the `Fixes #` section.

**Comment by vladimirvivien** [comment]
Please squash commits unless there is a need to separate commits.

**Comment by vladimirvivien** [comment]
@subramanian-neelakantan 
Do you have plans to add unit test files ?

**Comment by subramanian-neelakantan** [comment]
> /assign vladimirvivien

Thanks for reviewing this PR.

> @subramanian-neelakantan
> Since this is a bug, please create an issue explaining the problem. Then refer this issue in the `Fixes #` section.

Done. Created issue #73301 

> Please squash commits unless there is a need to separate commits.

Done.


> @subramanian-neelakantan
> Do you have plans to add unit test files ?

e2e tests have been added to cover this functionality and have passed. It is part of this PR #72732

**Comment by subramanian-neelakantan** [comment]
@andrewsykim I have updated the logic in admission controller after the merge of PR 
https://github.com/kubernetes/kubernetes/pull/73102. Could you take a look and approve?

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 4]
```diff
@@ -1418,3 +1436,258 @@ func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.Node
...
+		if zone.Region != "" && containsAll(dsHosts, zoneHosts) {
+			dsZones = append(dsZones, zone)
+		}
+	}
+	klog.V(4).Infof("Datastore %s belongs to zones: %v", datastore, dsZones)
+	return dsZones, nil
+}
+
+// GetZoneToHosts returns a map of 'zone' -> 'list of hosts in that zone' in given VC
+func (vs *VSphere) GetZoneToHosts(ctx context.Context, vsi *VSphereInstance) (map[cloudprovider.Zone][]vmwaretypes.ManagedObjectReference, error) {
```
Comment (vladimirvivien) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: @subramanian-neelakantan 
There is a lot going on here.  I would feel better if this was unit tested.

**Comment by vladimirvivien** [comment]
@subramanian-neelakantan 
This PR looks better. I left a comment above about a complex method that should be unit tested.

**Comment by andrewsykim** [comment]
Should also be easy to add a test case for vSphere volumes for this admission controller, see https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/persistentvolume/label/admission_test.go#L64-L686

**Comment by frapposelli** [comment]
@subramanian-neelakantan PTAL

**Comment by subramanian-neelakantan** [comment]
@vladimirvivien @andrewsykim @frapposelli I have updated the PR with unit tests to cover the complex logic in vsphere.go and additional test cases in the admission_test for vsphere. PTAL.

**Comment by subramanian-neelakantan** [comment]
kops is failing due to #73444

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 4 above)
Comment (subramanian-neelakantan) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: I have added unit tests to cover this now. Can you kindly take another look?

**Comment by subramanian-neelakantan** [comment]
@andrewsykim, @vladimirvivien, @frapposelli This change is now ready for a review and approval. PTAL

**Comment by vladimirvivien** [comment]
@subramanian-neelakantan thanks for the update.

/lgtm

File: pkg/volume/vsphere_volume/vsphere_volume.go [hunk 5]
```diff
@@ -384,7 +384,7 @@ func (v *vsphereVolumeProvisioner) Provision(selectedNode *v1.Node, allowedTopol
 	pv := &v1.PersistentVolume{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:   v.options.PVName,
-			Labels: map[string]string{},
+			Labels: volSpec.Labels,
```
Comment (andrewsykim) [inline pkg/volume/vsphere_volume/vsphere_volume.go]: The PVL admission controller handles labelling/node affinity rules for manually created volumes. But for dynamically created volumes, we still have to apply the volume's node affinity rules here as part of `Provision()`.

See [GCE PD](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gcepd/gce_pd.go#L539-L565) and [AWS EBS](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/awsebs/aws_ebs.go#L573-L588) for examples.

Comment (subramanian-neelakantan) [reply pkg/volume/vsphere_volume/vsphere_volume.go]: The PVL admission control seems to be handling the nodeAffinity for dynamic volumes as well. Still I will add this logic to the Provision() method in vsphere_volume just to be consistent with other volume plugins. The latest commit 9d6eb69 has this addressed. @andrewsykim PTAL.

**Comment by andrewsykim** [comment]
It's not ideal to add more providers to the (deprecated) PVL admission controller, but with support for Initailzers removed (https://github.com/kubernetes/kubernetes/pull/72972), we've had to backport some providers into the admission controller to account for lack of out-of-tree support (the out-of-tree PV labelling controller relied on Initializers). We've had to do the same for Cinder PVs as well (https://github.com/kubernetes/kubernetes/pull/73102). We're working through a better out-of-tree alternative in https://github.com/kubernetes/kubernetes/issues/73319 

/approve

**Comment by subramanian-neelakantan** [comment]
@frapposelli  I apologize, but had to push one more commit in order to rebase to upstream/master and resolve a conflict in pkg/volume/vsphere_volume/BUILD file. Could you kindly mark lgtm once again?

@SandeepPissay could you kindly take a look and approve?

**Comment by vladimirvivien** [comment]
@subramanian-neelakantan PR needs rebase.

**Comment by subramanian-neelakantan** [comment]
Waiting on @SandeepPissay approval.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 3 above)
Comment (SandeepPissay) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: > creates a new instance of a cloudprovider to fetch the zone information each time. So the cloudprovider cannot cache the zone information for a datastore.

We can avoid recomputing the zone info per datastore by caching this in global variable(a map?) if you are getting new instance of a cloud provider every time. 

Another concern I have is that the GetVolumeLabels is a heavy weight operation that does multiple remote calls to VC.  If zones feature is not used, it looks like this method is invoked irrespective of whether the zones is enabled or not. Am I right? If that is the case, could we optimize it such that if zones functionality is not enabled, then skip computing volume labels?

**Comment by SandeepPissay** [comment]
The main concern I have with this PR is performance - 
- datastore to zone/region mapping is static but we seem to be computing the zone for the same datastore for every volume created on it.
- computing volume labels logic is time consuming as it involves multiple remote calls. This is done even if zones feature is not enabled in k8s cluster.

**Comment by subramanian-neelakantan** [comment]
@SandeepPissay Thanks for the review and both good suggestions.

> The main concern I have with this PR is performance -
> 
> * datastore to zone/region mapping is static but we seem to be computing the zone for the same datastore for every volume created on it.

Let us track this performance improvement as a separate task since it fits in easily as an incremental change. Here is the issue I just created to track this - https://github.com/kubernetes/kubernetes/issues/73955

> * computing volume labels logic is time consuming as it involves multiple remote calls. This is done even if zones feature is not enabled in k8s cluster.

There is no way to disable the 'zones' feature in a k8s cluster AFAICT. The 'Labels' section of the vsphere.conf should not be treated as a feature gate. It only over-rides the tag category name used for zone and region.

**Comment by SandeepPissay** [comment]
@subramanian-neelakantan I would like to see the content of vsphere.conf with this file? Could you paste it here? 

I'm wondering if we can have a flag in vSphere.conf to explicitly enable multi zones support in VCP.

**Comment by frapposelli** [comment]
My understanding is that zones support is enabled only when the zones parameters are set in `vsphere.conf` otherwise support is disabled.

**Comment by SandeepPissay** [comment]
If zones feature is enabled by explicitly setting zones params in vsphere.conf, then VCP/vsphere volume plugin should not be invoking tag service APIs at all if the zones param is not present(zones features disabled). But that is not the case in this PR, which is what I'm highlighting here.

If there is a way to determine if zones feature is disabled, I would suggest to skip tag service API calls into Virtual Center. Else, this PR will introduce performance regression.

**Comment by subramanian-neelakantan** [comment]
With this change, zone support is enabled if the vCenter inventory (Datacenter/Cluster/Host) is tagged with zone/region tags. VCP looks for tags with the category name "zone" and "region" to identify this. The vsphere.conf is only used to over-ride the default category names of "zone" and "region". So Tag API calls are used to determine whether zones support is enabled or not.

My opinion is that we do not need another back door feature enablement (through vsphere.conf) for the zones support. It should be present once VCP has implemented this k8s feature. The performance implication of such a feature support, if any, should be addressed separately (see issue https://github.com/kubernetes/kubernetes/issues/73955).

OTOH, if we feel strongly against supporting zone feature by default in VCP, I can revert the part of code that makes "zone" and "region" as the default category names. That way, zones will be discovered from VC only when the vsphere.conf has such an entry:
[Labels]
    zone = <vc_zone_tag_category_name>
    region = <vc_region_tag_category_name>

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 6]
```diff
@@ -1426,3 +1444,258 @@ func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.Node
...
+			fds = make([]string, 0)
+		}
+		regionToZones[zone.Region] = append(fds, zone.FailureDomain)
+	}
+
+	// Join all fds in same region and return Zone instances
+	collapsedZones := make([]cloudprovider.Zone, 0)
+	for region, fds := range regionToZones {
+		fdSet := sets.NewString(fds...)
+		appendedZone := volumeutil.ZonesSetToLabelValue(fdSet)
```
Comment (andrewsykim) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: You may have to rebase this PR, a bunch of the volume zone helper methods were moved from `pkg/volume/util` to `k8s.io/cloud-provider/volume/helpers`.

**Comment by SandeepPissay** [comment]
Most of VMware customers are not using zones and may not need this feature, so this change will introduce performance regression for them which I'm suggesting to avoid. This can be a simple flag that gets set during VCP init.

**Comment by dougm** [comment]
We chose to make zones/tags optional to avoid breaking existing user configs on upgrades of K8s.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 6 above)
Comment (subramanian-neelakantan) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: Thanks for the notice. Rebased.

**Comment by subramanian-neelakantan** [comment]
> Most of VMware customers are not using zones and may not need this feature, so this change will introduce performance regression for them which I'm suggesting to avoid. This can be a simple flag that gets set during VCP init.

Since I have not heard any objections to this approach, I have made this change in commit ba9a9cf to treat the _"zone and region [Labels]"_ entries in vsphere.conf file to indicate whether zone support is enabled for VCP. When these entries are not present in vsphere.conf, VCP avoid zone based computation. In fact, it avoids discovering the tags from VC entirely.

@SandeepPissay PTAL.

**Comment by frapposelli** [comment]
thanks @dougm for the context

@subramanian-neelakantan the approach you took with the latest commit seems to follow the original intention of the Zones feature so

/lgtm

leaving approval to @SandeepPissay

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 7]
```diff
@@ -1438,3 +1446,269 @@ func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.Node
...
+	if pv.Spec.VsphereVolume == nil {
+		return nil, nil
+	}
+	return vs.GetVolumeLabels(pv.Spec.VsphereVolume.VolumePath)
+}
+
+// GetVolumeLabels returns the well known zone and region labels for given volume
+func (vs *VSphere) GetVolumeLabels(volumePath string) (map[string]string, error) {
+	// Create context
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (SandeepPissay) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Move this code to after zones check as this is redundant if zones is not enabled.

**Review by SandeepPissay — APPROVED** [review]
/lgtm

**Comment by vladimirvivien** [comment]
@SandeepPissay don't forget to mark this PR as `/approved` so it'll get merged.

**Comment by vladimirvivien** [comment]
Will this be cherry picked to an earlier version?

**Pull request merged**